[FLINK-17861][task][checkpointing] Split channel state handles sent to JM

rkhachatryan opened the pull request:
## What is the purpose of the change

If Unaligned checkpoints are enabled, channel state is written as state handles. Each channel has a handle and each such handle references the same underlying `streamStateHandle` (this is done to have a single file per subtask).
But, if the state is less then `state.backend.fs.memory-threshold`, the data is sent directly to JM as a byteStreamHandle. This causes each channel state handle to hold the whole subtask state. 

This PR solves this by extracting relevant potions of the underlying handles if they are `byteStreamHandle`s.

Another approach would be to move the underlying state handle one level up (`OperatorSubtaskState`) and not store references to it in `channelStateHandles`. It would be more effective (less data duplication) but also more error-prone (implicit structure), less flexible (re-scaling), and require more changes.

## Verifying this change

This change is already covered by existing tests, such as:
 - `ChannelStateCheckpointWriterTest.testRecordingOffsets` (the test was corrected)
 - `ChannelPersistenceITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriter.java:
```diff
@@ -180,17 +177,33 @@ private void doComplete(boolean precondition, RunnableWithException complete, Ru
+			StreamStateHandle underlying,
+			TriFunction<I, StreamStateHandle, List<Long>, H> buildHandle,
+			Map.Entry<I, List<Long>> e) throws IOException {
+		if (underlying instanceof ByteStreamStateHandle) {
+			ByteStreamStateHandle byteHandle = (ByteStreamStateHandle) underlying;
+			return buildHandle.apply(
+				e.getKey(),
+				new ByteStreamStateHandle(randomUUID().toString(), serializer.extractAndMerge(byteHandle.getData(), e.getValue())),
+				singletonList(serializer.getHeaderLength()));
+		} else {
```
It doesn't look like a generic/error prone solution. It looks like something should be extracted to the `StreamStateHandle`. `boolean StreamStateHandle#isDirectlyHoldingData()`? Or we are missing some kind of different abstraction of shared state handles.

Why is this issue new for spilled channel state? What's different for operators state?

rkhachatryan replied:
Added `StreamStateHandle.asBytesIfInMemory` that returns `Optional<byte[]>` (better names are welcomed ;)

> Why is this issue new for spilled channel state? What's different for operators state?

Operators don't share state handles with each other so they don't have this problem.

pnowojski replied:
I think it's better now, but I still can not put a finger on a feeling that something might not be entirely correct. 

Why do we have multiple handles per channel, instead of a single handle that has multiple channels?

rkhachatryan replied:
There is only one `InputChannelStateHandle` per channel.
Putting multiple channels could make recovery (rescaling) more difficult.

pnowojski replied:
After an offline discussion, we decided to keep it for now as it is. Could you create a JIRA ticket to revisit this issue later (linked to how to handle rescaling) and linked it as `TODO` here?

rkhachatryan replied:
Sure. Created a JIRA ticket: https://issues.apache.org/jira/browse/FLINK-17972

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriter.java:
```diff
@@ -25,8 +25,10 @@
 import org.apache.flink.runtime.state.InputChannelStateHandle;
```
> some more elaborate explanation what was the underlying problem

I've updated PR description since then, does it look good to you now?

pnowojski replied:
Thanks, now it's better :) Could you copy this updated description  to commit message and jira ticket?

rkhachatryan replied:
Done.

rkhachatryan commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriterTest.java:
```diff
@@ -149,8 +152,8 @@ public void testRecordingOffsets() throws Exception {
 	private void write(ChannelStateCheckpointWriter writer, InputChannelInfo channelInfo, byte[] data) throws Exception {
-		NetworkBuffer buffer = new NetworkBuffer(HeapMemorySegment.FACTORY.allocateUnpooledSegment(data.length, null), FreeingBufferRecycler.INSTANCE);
-		buffer.setBytes(0, data);
+		MemorySegment segment = wrap(data);
+		NetworkBuffer buffer = new NetworkBuffer(segment, FreeingBufferRecycler.INSTANCE, Buffer.DataType.DATA_BUFFER, segment.size());
```
Writer position of `NetworkBuffer` wasn't updated.This change updates it by passing `segment.size()` to constructor.

Updated commit message.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateCheckpointWriter.java:
```diff
@@ -180,17 +177,33 @@ private void doComplete(boolean precondition, RunnableWithException complete, Ru
 			CompletableFuture<Collection<H>> future,
 			Map<I, List<Long>> offsets,
-			BiFunction<I, List<Long>, H> buildHandle) {
+			TriFunction<I, StreamStateHandle, List<Long>, H> buildHandle) throws IOException {
```
Totally agree.
Extracted `HandleFactory` interface and replaced `Map.Entry` with two arguments.

rkhachatryan commented:
Thanks for the feedback @pnowojski ,

I've addressed the issues (except [this one](https://github.com/apache/flink/pull/12292#discussion_r429887132)).

Answering your question:
> Could you elaborate a bit more? What's the alternative? How would it avoid more data duplication? Are we still duplicating data with this PR?

Current structure is the following (this PR doesn't change it):
```
Each subtask reports to JM TaskStateSnapshot, 
    each with zero ore more OperatorSubtaskState,
        each with zero or more InputChannelStateHandle and ResultSubpartitionStateHandle
            each referencing an underlying StreamStateHandle
```
The underlying `StreamStateHandle` duplicates filename (`ByteStreamStateHandle` has it too at least because of `equals/hashcode` I guess).

An alternative would be something like 
```
Each subtask reports to JM TaskStateSnapshot, 
    each with zero ore more OperatorSubtaskState,
        each with zero or one StreamStateHandle for channel state handles
            each with zero or more InputChannelStateHandle and ResultSubpartitionStateHandle
```

pnowojski commented:
My questions probing in the direction of single channel state handle seems to be going in a very similar direction as the alternate solution that you proposed: moving the shared `StreamStateHandle` one level up, but also additionally encapsulating them into something, for example:

```
Each subtask reports to JM TaskStateSnapshot, 
    each with zero ore more OperatorSubtaskState,
        each with zero or one InFlightDataStateHandle
            each with one StreamStateHandle
            each with zero or more InputChannelStateHandle and ResultSubpartitionStateHandle
```

rkhachatryan commented:
@flinkbot run azure

pnowojski commented:
Merging from https://github.com/apache/flink/pull/12362 . Build was successful there, here e2e timed out because of getting artefacts from mvn cache took very long time
